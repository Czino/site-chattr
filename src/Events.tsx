import { useSubscribe } from 'nostr-hooks'
import { useMemo } from 'react'
import { Event } from './components/Event'
import { getDomain } from './helpers/getDomain'

const opts = {
    closeOnEose: false,
}
const relays = [
    'wss://nostr-pub.wellorder.net/',
    'wss://nostr.bitcoiner.social',
    'wss://relay.damus.io/',
]
type Props = {
    url: string
}

const GENERIC_COMMENT_KIND = 1111
export const Events = ({ url }: Props) => {
    const filters = useMemo(
        () => [
            {
                kinds: [GENERIC_COMMENT_KIND],
                '#S': [`r:${url}`],
                '#K': [`r:${getDomain(url)}`],
            },
        ],
        [url],
    )

    const { events, eose } = useSubscribe({
        filters,
        opts,
        relays,
        enabled: !!url,
        fetchProfiles: true,
    })

    if (!eose) return <div className="text-center text-purple-300">Fetching messages...</div>
    if (events.length === 0) return <div className="text-center text-purple-300">No messages yet</div>

    return (
        <ul className="flex flex-col-reverse w-full gap-4">
            {events.map((event) => (
                <Event key={event.id} event={event} />
            ))}
        </ul>
    )
}
