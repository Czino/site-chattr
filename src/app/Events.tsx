'use client'
import { useSubscribe } from 'nostr-hooks'
import { useMemo } from 'react'

type Props = {
    currentDomain: string
    currentURL: string
}
export const Events = ({ currentDomain, currentURL }: Props) => {
    const filters = useMemo(
        () => [
            {
                kinds: [1111],
                '#S': [`r:${currentURL}`],
                '#K': [`r:${currentDomain}`],
            },
        ],
        [],
    )
    const { events, eose } = useSubscribe({ filters, fetchProfiles: true })

    if (!eose) return <div>Fetching messages...</div>

    return (
        <ul className="grid w-full gap-4">
            {events.map((event) => (
                <li key={event.id} className="w-full bg-slate-50 rounded p-4 overflow-hidden">
                    <p>{event.author.profile?.displayName}</p>
                    <p>{event.content}</p>
                    <p>{JSON.stringify(event.tags)}</p>
                    <p>{event.id}</p>
                </li>
            ))}
        </ul>
    )
}
