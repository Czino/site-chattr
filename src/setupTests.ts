import '@testing-library/jest-dom'
import { toMatchDiffSnapshot } from 'snapshot-diff'
import { TextDecoder, TextEncoder } from 'util'
expect.extend({ toMatchDiffSnapshot })
Object.assign(global, { TextDecoder, TextEncoder })

export const testPubkey = 'c446c97e0b3a2f5e1d33ba40664fa77b03b0afcba9684daf4f7925ae8f5c6fe0'
export const mockLoginFromLocalStorage = jest.fn().mockImplementation(({ onSuccess }) => {
    if (onSuccess) onSuccess()
})
export const mockLoginWithSecretKey = jest.fn().mockImplementation(({ onSuccess }) => {
    if (onSuccess) onSuccess()
})
export const mockLoginWithExtension = jest.fn().mockImplementation(({ onSuccess }) => {
    if (onSuccess) onSuccess()
})
export const mockLogout = jest.fn()
export const mockUseSubscribe = jest.fn().mockReturnValue({
    App: [],
    eose: false,
})
jest.mock('nostr-hooks', () => ({
    useNostrHooks: jest.fn(),
    useLogin: () => ({
        loginFromLocalStorage: mockLoginFromLocalStorage,
        loginWithExtension: mockLoginWithExtension,
        loginWithSecretKey: mockLoginWithSecretKey,
        logout: mockLogout,
    }),
    useActiveUser: () => ({
        activeUser: { pubkey: testPubkey },
    }),
    useSubscribe: (...args: unknown[]) => mockUseSubscribe(...args),
    useNewEvent: () => ({ createNewEvent: jest.fn() }),
}))
