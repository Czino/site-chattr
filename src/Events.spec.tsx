import NDK, { NDKEvent } from '@nostr-dev-kit/ndk'
import { render } from '@testing-library/react'
import { event1, event2 } from '../test/data/eventData'
import { Events } from './Events'
import { mockUseSubscribe } from './setupTests'

describe('Events', () => {
    const url = 'https://localhost:3000/'
    it('should subscribe with the correct filters', () => {
        render(<Events url={url} />)
        expect(mockUseSubscribe).toHaveBeenCalledWith({
            enabled: true,
            fetchProfiles: true,
            filters: [
                {
                    '#r': ['https://localhost:3000/'],
                    kinds: [1],
                },
                {
                    '#K': ['https://localhost:3000'],
                    '#I': ['https://localhost:3000/'],
                    kinds: [1111],
                },
            ],
            opts: {
                closeOnEose: false,
            },
            relays: ['wss://nostr-pub.wellorder.net/', 'wss://nostr.bitcoiner.social', 'wss://relay.damus.io/'],
        })
    })
    it('should render correctly while fetching', () => {
        const base = render(<Events url={url} />).asFragment()
        expect(base).toMatchSnapshot()
    })
    it('should render correctly when no messages could be fetched', () => {
        mockUseSubscribe.mockReturnValue({
            events: [],
            eose: true,
        })
        const base = render(<Events url={url} />).asFragment()
        expect(base).toMatchSnapshot()
    })
    it('should render messages', () => {
        const ndk = new NDK()

        mockUseSubscribe.mockReturnValue({
            events: [new NDKEvent(ndk, event1), new NDKEvent(ndk, event2)],
            eose: true,
        })
        const base = render(<Events url={url} />).asFragment()
        expect(base).toMatchSnapshot()
    })
})
