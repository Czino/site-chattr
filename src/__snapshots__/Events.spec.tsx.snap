// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events should render correctly when no messages could be fetched 1`] = `
<DocumentFragment>
  <div
    class="text-center text-purple-300"
  >
    No messages yet
  </div>
</DocumentFragment>
`;

exports[`Events should render correctly while fetching 1`] = `
<DocumentFragment>
  <div
    class="text-center text-purple-300"
  >
    Fetching messages...
  </div>
</DocumentFragment>
`;

exports[`Events should render messages 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col-reverse gap-4 w-full"
  >
    <li
      class="w-full rounded p-4 overflow-hidden flex flex-col gap-1 bg-purple-100"
    >
      <a
        class="flex gap-1 items-center leading-5"
        href="https://njump.me/npub1xunyx3crpkgem3nntqudrrg5pnp329q27p5yz59ggccc7lq604vsan554l"
        rel="noreferrer"
        target="_blank"
      >
        <p
          class="text-xs text-purple-400"
        >
          9/17/2024, 2:47:01 PM
        </p>
      </a>
      <p>
        Home sweet home
      </p>
    </li>
    <li
      class="w-full rounded p-4 overflow-hidden flex flex-col gap-1 bg-highlight"
    >
      <a
        class="flex gap-1 items-center leading-5"
        href="https://njump.me/npub1c3rvjlst8gh4u8fnhfqxvna80vpmpt7t495ymt600yj6ar6udlsq60swvh"
        rel="noreferrer"
        target="_blank"
      >
        <p
          class="text-xs text-purple-400"
        >
          9/18/2024, 10:04:39 AM
        </p>
      </a>
      <p>
        I develop almost all my apps here!
      </p>
    </li>
  </ul>
</DocumentFragment>
`;
